### PR TITLE
Node Affinity and Node Anti-Affinity enhancement related changes.

### PR DESCRIPTION
**Pull request by shivramsrivastava**
Refer #51 

Pending:

- [ ] Node Affinity E2E test cases.

### CODE REVIEW

**Comment by m1093782566** [comment]
@shivramsrivastava 

A quick question. I find you have changed some existing `*.pb.go`. I am wondering why since you just added a new protobuf file. What command did you use to generate `*.pb.go`? 

```
protoc --go_out=. *.proto
```

?

**Comment by nikita15p** [comment]
@m1093782566 
protoc --go_out=,import_path=$GOPATH/src/github.com/kubernetes-sigs/poseidon/pkg/firmament,plugins=grpc:. *.proto is used for generating *.pb.go

**Comment by m1093782566** [comment]
@nikita15p 

Thanks for your reply, I am fine with it. I just want a clarification :)

**Comment by m1093782566** [comment]
LGTM overall except some nits.

File: pkg/firmament/task_desc.proto [hunk 1]
```diff
@@ -101,4 +102,6 @@ message TaskDescriptor {
   repeated Label labels = 32;
   // Resource label selectors
   repeated LabelSelector label_selectors = 33;
+  //Affinity
```
[the remark this answers is not in the document]
Comment (nikita15p) [reply pkg/firmament/task_desc.proto]: Yes, the kubernetes also follow same naming convention ie Affinity being the superset of Node Affinity, Pod Affinity and Pod AntiAffinity.

File: pkg/k8sclient/podwatcher.go [hunk 2]
```diff
@@ -146,6 +147,90 @@ func (pw *PodWatcher) getCPUMemRequest(pod *v1.Pod) (int64, int64) {
 	return cpuReq, memReq
 }
 
+func (pw *PodWatcher) getNodeSelectorTerm(pod *v1.Pod) []NodeSelectorTerm {
+
+	var nodeSelTerm []NodeSelectorTerm
+	if pod.Spec.Affinity != nil {
+		if pod.Spec.Affinity.NodeAffinity != nil {
+			if pod.Spec.Affinity.NodeAffinity.RequiredDuringSchedulingIgnoredDuringExecution != nil {
+				err := copier.Copy(&nodeSelTerm, pod.Spec.Affinity.NodeAffinity.RequiredDuringSchedulingIgnoredDuringExecution.NodeSelectorTerms)
```
[the remark this answers is not in the document]
Comment (nikita15p) [reply pkg/k8sclient/podwatcher.go]: Yes it does.

File: pkg/k8sclient/podwatcher.go [hunk 3]
```diff
@@ -146,6 +147,90 @@ func (pw *PodWatcher) getCPUMemRequest(pod *v1.Pod) (int64, int64) {
...
+				if err != nil {
+					glog.Info("NodeSelectorTerm Data could not be copied")
```
[the remark this answers is not in the document]
Comment (nikita15p) [reply pkg/k8sclient/podwatcher.go]: Done

File: pkg/k8sclient/podwatcher.go [hunk 4]
```diff
@@ -146,6 +147,90 @@ func (pw *PodWatcher) getCPUMemRequest(pod *v1.Pod) (int64, int64) {
...
+}
+
+func (pw *PodWatcher) getPreferredSchedulingTerm(pod *v1.Pod) []PreferredSchedulingTerm {
+	var prefSchTerm []PreferredSchedulingTerm
+	if pod.Spec.Affinity != nil {
+		if pod.Spec.Affinity.NodeAffinity != nil {
+
+			err := copier.Copy(&prefSchTerm, pod.Spec.Affinity.NodeAffinity.PreferredDuringSchedulingIgnoredDuringExecution)
+			if err != nil {
+				glog.Info("PreferredSchedulingTerm Data could not be copied")
```
[the remark this answers is not in the document]
Comment (nikita15p) [reply pkg/k8sclient/podwatcher.go]: Done.

File: pkg/firmament/pod_anti_affinity.proto [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+message LabelSelectorAntiAff{
+
+    MatchLabelsAntiAff matchLabels = 1;
+
+  repeated LabelSelectorRequirementAntiAff matchExpressions = 2;
+}
+
+message PodAffinityTermAntiAff {
+
+    LabelSelectorAntiAff labelSelector = 1;
```
[the remark this answers is not in the document]
Comment (nikita15p) [reply pkg/firmament/pod_anti_affinity.proto]: Done.

File: pkg/firmament/pod_anti_affinity.proto [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+  string key = 1;
+
+  string operator = 2;
+
+  repeated string values = 3;
+}
+
+message LabelSelectorAntiAff{
+
+    MatchLabelsAntiAff matchLabels = 1;
```
[the remark this answers is not in the document]
Comment (nikita15p) [reply pkg/firmament/pod_anti_affinity.proto]: Done.

**Pull request merged**